### PR TITLE
Code review for employee service extension with holiday request service and mail

### PR DESCRIPTION
**Pull request by agyen**
our added files are RequestTO, RequestController, Email and GmailService

### CODE REVIEW

File: db/employmentDb.sql [hunk 1]
```diff
@@ -1,51 +1,64 @@
-
 CREATE TABLE employee (
-
 	employee_id serial primary key,
 	employee_firstName text not null,
 	employee_lastName text not null,
 	employee_phoneNumber text not null,
 	employee_email text not null,
 	employee_address text not null,
+	employee_role text not null,
```
Comment (sammoorhouse) [inline db/employmentDb.sql:8]: If you make this change to an existing database, what will happen? 

you've added a new text column which cannot be null. What will be the value of this column in all the existing rows?

File: db/employmentDb.sql [hunk 2]
```diff
@@ -1,51 +1,64 @@
...
-);
-
-CREATE TABLE project (
-
-	project_id serial primary key,
-	project_name text not null,
-	project_description text not null,
-	project_start_date DATE not null,
-	project_date_date DATE not null
+	employee_gender text not null,
```
Comment (sammoorhouse) [inline db/employmentDb.sql:12]: why do we need to know this?

File: db/employmentDb.sql [hunk 3]
```diff
@@ -1,51 +1,64 @@
...
-);
-
-CREATE TABLE assignedproject (
-
-	assignedproject_id serial primary key,
+CREATE TABLE assignproject (
+    assignproject_id serial primary key,
 	employee_id INTEGER,
-	project_id INTEGER
+	project_id INTEGER,
```
Comment (sammoorhouse) [inline db/employmentDb.sql:19]: what does this have to do with holiday calendars?!

File: db/employmentDb.sql [hunk 4]
```diff
@@ -1,51 +1,64 @@
...
+	project_id int references projects (project_id),
+	project_hours int,
+	primary key (project_date,project_id)
 );
 
-CREATE TABLE employeetech (
-	techproject_id serial primary key,
-	tech_id INTEGER,
-	employee_id INTEGER
+create table LoggedSick (
```
Comment (sammoorhouse) [inline db/employmentDb.sql:33]: what does this have to do with holiday calendars?!

File: db/holiday.sql [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+CREATE TABLE requests(
+    request_id serial primary key NOT NULL,
+    requester_id int references employee(employee_id) NOT NULL,
+    request_start_date date NOT NULL,
+    request_report_date date NOT NULL,
+    request_status_id int references request_status(request_status_id) NOT NULL default 1
+);
+
+--- -------------------------------------------------------------------------------------------- ---
+                        --- Dumping table for employees ---
```
Comment (sammoorhouse) [inline db/holiday.sql:24]: what does this comment mean?

File: db/holiday.sql [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+
+INSERT INTO request_status(req_status) VALUES ('PENDING');
+INSERT INTO request_status(req_status) VALUES ('DECLINED');
+INSERT INTO request_status(req_status) VALUES ('APPROVED');
+
+--- -------------------------------------------------------------------------------------------- ---
+                        --- Dumping table for requests ---
+--- -------------------------------------------------------------------------------------------- ---
+
+INSERT INTO requests(requester_id, request_start_date, request_report_date, request_status_id) values(1,'2020-02-20', '2020-02-25', 2);
```
Comment (sammoorhouse) [inline db/holiday.sql:35]: so if you want to insert sample data you should have a separate script that does that. If we run this in production we'll get all the new tables plus a load of nonsense data we then have to clean up

File: src/main/java/io/turntabl/employementprofilingsystem/Controllers/AuthController.java [hunk 7]
```diff
@@ -0,0 +1,66 @@
...
+        try{
+            request.put("employee_email", email);
+
+            List<String> requiredParams = Arrays.asList(
+                    "employee_email"
+            );
+            Map<String, Object> valid = parsor.validate_params(request, requiredParams);
+            if (valid.get("code").equals("00")) {
+                List<Employee>  employee = jdbcTemplate.query(
+                        "select * from employee where employee_email = ?",
```
Comment (sammoorhouse) [inline src/main/java/io/turntabl/employementprofilingsystem/Controllers/AuthController.java:41]: don't select * - what if someone adds a column (which is exactly what you're doing!) - all the existing queries will break. Instead, make your code more resilient by naming the columns in the select clause.

File: src/main/java/io/turntabl/employementprofilingsystem/Controllers/AuthController.java [hunk 8]
```diff
@@ -0,0 +1,66 @@
...
+                        new Object[]{email},
+                        BeanPropertyRowMapper.newInstance(Employee.class)
+                );
+                if (!employee.isEmpty()){
+                    response.put("code","00");
+                    response.put("msg","Data retrieved successfully");
+                    response.put("data",employee.get(0));
+                }else{
+                    response.put("code","00");
```
Comment (sammoorhouse) [inline src/main/java/io/turntabl/employementprofilingsystem/Controllers/AuthController.java:50]: what is code 00?

File: src/main/java/io/turntabl/employementprofilingsystem/Controllers/EmployeeController.java [hunk 9]
```diff
@@ -0,0 +1,308 @@
...
+
+            List<String> requiredParams = Arrays.asList(
+                    "employee_email",
+                    "employee_role"
+            );
+            Map<String, Object> result = parsor.validate_params(request,requiredParams);
+            if (result.get("code").equals("00")){
+
+                if (this.checkExistingEmployee(requestData.getEmployee_email())){
+                    response.put("code","01");
```
Comment (sammoorhouse) [inline src/main/java/io/turntabl/employementprofilingsystem/Controllers/EmployeeController.java:57]: what are these codes?

File: src/main/java/io/turntabl/employementprofilingsystem/Controllers/RequestController.java [hunk 10]
```diff
@@ -0,0 +1,74 @@
...
+    public void makeARequest(@RequestBody RequestTO request) {
+        jdbcTemplate.update("insert into requests(requester_id, request_start_date, request_report_date) values(?,?,?)",
+                request.getRequester_id(), request.getRequest_start_date(), request.getRequest_report_date());
+
+        SimpleDateFormat DateFor = new SimpleDateFormat("E, dd MMMM yyyy");
+        String startDate = DateFor.format(request.getRequest_start_date());
+        String endDate = DateFor.format(request.getRequest_report_date());
+
+        try {
+            Email.requestMessage("isaac.agyen@turntabl.io", request.getFrom(), "Holiday request", startDate, endDate, request.getRequester_name());
```
Comment (sammoorhouse) [inline src/main/java/io/turntabl/employementprofilingsystem/Controllers/RequestController.java:33]: this email shouldn't be hardcoded

**Review by sammoorhouse — COMMENTED** [review]
So two things straight off - 1. don't call branches "code review". Name the branch for the feature. Similarly, name the pull request after the contents. 2. The PR isn't "code review for employee service extension...", it's "extend employee service with remaining holiday details"

Finally, this PR doesn't seem to be just holiday service code - there's a bunch of other changes in here. Can you create different PRs for different changes?

**Review by sammoorhouse — CHANGES_REQUESTED** [review]
So two things straight off - 1. don't call branches "code review". Name the branch for the feature. Similarly, name the pull request after the contents. 2. The PR isn't "code review for employee service extension...", it's "extend employee service with remaining holiday details"

Finally, this PR doesn't seem to be just holiday service code - there's a bunch of other changes in here. Can you create different PRs for different changes?

File: db/employmentDb.sql (hunk 2 above)
Comment (agyen) [reply db/employmentDb.sql:12]: we won't need this, is for tpms

File: db/employmentDb.sql (hunk 3 above)
Comment (agyen) [reply db/employmentDb.sql:19]: yeah It has nothing to do with it. its for tpms :)

File: db/holiday.sql (hunk 5 above)
Comment (agyen) [reply db/holiday.sql:24]: It was a dummy data for testing, we'll dump it when pushing to production
